### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by CreaTiVeUK**
Module 2 - Optional changes
Module 3 - Completed
Module 4 - Part of it

### CODE REVIEW

File: todo_app/app.py [hunk 1]
```diff
@@ -1,59 +1,66 @@
...
-app = Flask(__name__)
-app.config.from_object(Config)
 
+def create_app():
+    app = Flask(__name__)
+    app.config.from_object(Config())
+    
+    @app.route('/')
+    def index():
+        items = session.get_Items_and_Status()[0]
```
Comment (JackMead) [inline todo_app/app.py:14]: Given this function is only used here, we could change it to only return the items, and then we don't need to index it with `[0]`

File: todo_app/data/logs/app.log [hunk 2]
```diff
@@ -157,3 +157,141 @@
 ### 2022-03-02 02:49:23,086 ### appLog ### INFO ### Card due date: 07/12/2022 ###
```
Comment (JackMead) [inline todo_app/data/logs/app.log:157]: We chatted through removing this already, but just to add I think that's a good shout but also nice work on getting a log file set up!

File: todo_app/tests/test_unit_tests.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    list.append(Item(1, "Testing", "A card used for testing", "03/15/2022", "03/15/2022", "To Do", 12345))
+    list.append(Item(2, "Testing", "A card used for testing", "03/15/2022", "03/15/2022", "Doing", 12345))
+    list.append(Item(3, "Testing", "A card used for testing", "03/15/2022", "03/15/2022", "Done", 12345))
+    list.append(Item(4, "Testing", "A card used for testing", "03/15/2022", "03/15/2022", "Done", 12345))
+    item_view_model = ViewModel(list)
+    #Act
+    done_items = item_view_model.done_items
+    #Assert
+    assert 2 == len(done_items)
+    assert done_items[0].list_name == "Done"
```
Comment (JackMead) [inline todo_app/tests/test_unit_tests.py:43]: Nice; clear & well targeted unit tests here.

**Review by JackMead — APPROVED** [review]
This looks great, I was easily able to follow your instructions both to run the app with the updates and to run the tests without issues. I've left a few optional suggestions below, but very happy for you to merge this when you're ready

File: README.md [hunk 4]
```diff
@@ -54,6 +53,57 @@ You should see output similar to the following:
...
+```bash
+$ ansible-playbook --ask-vault-pass deploy_todoApp.yaml -i inventory
+```
+
+7. In your browser enter the managed node IP address at port 5000 to view the app i.e.:
+123.123.123.123:5000
+
+## Testing the App
+
+You can run the available tests pack, in development mode within the poetry environment by running the following from the project's root directory.
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply README.md:94]: Thanks Jack!

File: pyproject.toml [hunk 5]
```diff
@@ -9,6 +9,9 @@ python = "^3.7"
 flask = "^1.1.2"
 python-dotenv = "^0.14.0"
 requests = "^2.27.1"
+pytest = "^7.0.1"
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply pyproject.toml:12]: Thanks for the suggestion. Fixed.

File: todo_app/data/session_items_trello.py [hunk 6]
```diff
@@ -13,17 +37,96 @@ def get_Items_and_Status():
...
+        for card in list['cards']:         
             cards.append(Item(
-                card['id'], card['name'], card['desc'], due, list['name'], list['id']))
+                card['id'],
+                card['name'],
+                card['desc'],
+                card['due'] if card['due'] is None else date_parse(card['due']),
+                date_parse(card['dateLastActivity']),
+                list['name'],
+                list['id'])
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply todo_app/data/session_items_trello.py:54]: Thanks Jack.

File: todo_app/data/ViewModel.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            return items
+
+        @property
+        def done_items(self):
+            items = []
+            for item in self.items:
+                if item.list_name == "Done":
+                    items.append(item)
+            return items
+    """
```
[the remark this answers is not in the document]
Comment (CreaTiVeUK) [reply todo_app/data/ViewModel.py:52]: Removed. Thanks.

File: todo_app/app.py (hunk 1 above)
Comment (CreaTiVeUK) [reply todo_app/app.py:14]: Fixed. Thanks.

**Pull request merged**